### ISSUE TITLE
Next basePath not handled on signin page

### ISSUE THREAD

**Issue by marcoreni** [nextauthjs/next-auth#9425]
### Environment

  System:
    OS: macOS 14.2
    CPU: (10) arm64 Apple M1 Pro
    Memory: 149.20 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/Library/Caches/fnm_multishells/2567_1703023890015/bin/node
    Yarn: 1.22.19 - ~/.yarn/bin/yarn
    npm: 10.2.3 - ~/Library/Caches/fnm_multishells/2567_1703023890015/bin/npm
    pnpm: 8.12.1 - /opt/homebrew/bin/pnpm
    bun: 0.5.8 - ~/.bun/bin/bun
  Browsers:
    Chrome: 120.0.6099.109
    Safari: 17.2
  npmPackages:
    next: 14.0.4 => 14.0.4
    next-auth: 5.0.0-beta.4 => 5.0.0-beta.4
    react: 18.2.0 => 18.2.0

### Reproduction URL

https://github.com/marcoreni/next-auth-repro-base-path

### Describe the issue

My application has a basePath configured on "next" side. However, this is not fully handled properly on Next.

Session call work, redirection to signin page works, but the form action does not contain the 

### How to reproduce

1. Configure a `basePath` inside next.config.js
2. Configure the base path according to Next-Auth configuration (both NEXTAUTH_URL and `basePath` inside session provider)
3. Open the project and click on Signin. You will be redirected to /api/auth/signin instead of /{basePath}/api/auth/signin.

(NOTE: the repro repository has been configured with a base path of "/foo". If you open http://localhost:3000/foo/middleware-example , you will be redirected on the login page, and clicking on the Github button will yield the problem)

### Expected behavior

The form action URL should include the basePath.

**Comment by jenewland1999** [comment]
Behaviour is still present in in Next.js v14.1 and next-auth v5 beta.5

**Comment by balazsorban44** [comment]
The next.config.js basePath cannot be propagated to NextAuth.js, so you need to manually add this as a prop to `SessionProvider` and to the `NextAuthConfig`.

**Comment by jenewland1999** [comment]
@balazsorban44 According to the types being shipped in beta.5, `basePath` is not a valid option for `NextAuthConfig`. Also, if we use App Router, do we still need `SessionProvider` to wrap the entire app in the Root Layout?

![image](https://github.com/nextauthjs/next-auth/assets/34895363/c18b22ba-9dbf-4a00-a77e-793baf0d6816)

**Comment by balazsorban44** [comment]
Not shipped yet but it's here https://github.com/nextauthjs/next-auth/blob/b5d5f20206a08f9cb43aa0059d32963d4e90d088/packages/next-auth/src/lib/env.ts#L12


`SessionProvider` will likely be not needed in App Router at all. We'll document this soon.

**Comment by jenewland1999** [comment]
Okay, thank you. I'll keep an eye on the releases page for now. 😄

**Comment by hscstudio** [comment]
So, now, can I do for it? I must run Next 14 App router on sub directory, but fail becaus auth 5 not yet supported..

http://localhost:3000/admin

I have add in SessionProvider

<SessionProvider basePath="/admin/api/auth">

and

in AuthConfig

basePath:"/admin/api/auth"

but still error

**Comment by neoplastic** [comment]
@hscstudio I'm getting this same problem. Mine says in the logs 
`Cannot parse action at /api/auth/callback/nodemailer. Read more at https://errors.authjs.dev#unknownaction`

**Comment by jbogedahl-st** [comment]
> Not shipped yet but it's here
> 
> https://github.com/nextauthjs/next-auth/blob/b5d5f20206a08f9cb43aa0059d32963d4e90d088/packages/next-auth/src/lib/env.ts#L12
> 
> `SessionProvider` will likely be not needed in App Router at all. We'll document this soon.

I don't understand... from what I'm seeing, the client-side code completely ignores this.  If you look at the source where __NEXTAUTH is initialized, on the client side the process.env.{VARIABLES} are all undefined. But we don't try to use the config.basePath on the client-side.  So, here's the __NEXTAUTH initializer:

```export const __NEXTAUTH = {
    baseUrl: parseUrl(process.env.NEXTAUTH_URL ?? process.env.VERCEL_URL).origin,
    basePath: parseUrl(process.env.NEXTAUTH_URL).path,
    baseUrlServer: parseUrl(process.env.NEXTAUTH_URL_INTERNAL ??
        process.env.NEXTAUTH_URL ??
        process.env.VERCEL_URL).origin,
    basePathServer: parseUrl(process.env.NEXTAUTH_URL_INTERNAL ?? process.env.NEXTAUTH_URL).path,
    _lastSync: 0,
    _session: undefined,
    _getSession: () => { },
};```

and later, when we try to get the api URLs we ultimately call apiBaseUrl() which just does this:

```export function apiBaseUrl(__NEXTAUTH) {
    if (typeof window === "undefined") {
        // Return absolute path when called server side
        return `${__NEXTAUTH.baseUrlServer}${__NEXTAUTH.basePathServer}`;
    }
    // Return relative path when called client side
    return __NEXTAUTH.basePath;
}```

So bottom line question, how do I override the basePath on the client-side in authjs and next-auth/react api calls from the client?  I need to be able to call https://{host}/app/api/auth/[...nextauth]/route.ts

**Comment by mirkahromi** [comment]
> The next.config.js basePath cannot be propagated to NextAuth.js, so you need to manually add this as a prop to `SessionProvider` and to the `NextAuthConfig`.

Could you please elaborate a bit more how to set them up?

**Comment by nktnet1** [comment]
Collection of workarounds I've found:
1. [k3k8's solution](https://github.com/nextauthjs/next-auth/pull/10797#issuecomment-2303802441)
2. [ThanhHuuVu's solution](https://github.com/nextauthjs/next-auth/pull/10797#issuecomment-2253748338)
3. [andreynovikov's solution](https://github.com/nextauthjs/next-auth/discussions/12160)

**Comment by SrayasPMaatrum** [comment]
for v5 [next-auth/discussions/9274](https://github.com/nextauthjs/next-auth/discussions/9274) this works for me
